gtk4/src/builder_rust_scope/imp.rs

```diff
@@ -0,0 +1,120 @@
+
+use crate::prelude::*;
+use crate::subclass::prelude::*;
+use crate::Builder;
+use crate::BuilderClosureFlags;
+use crate::BuilderError;
+use crate::BuilderScope;
+use glib::translate::*;
+use glib::Closure;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+type Callback = dyn Fn(&[glib::Value]) -> Option<glib::Value>;
+
+#[derive(Default)]
+pub struct BuilderRustScope {
+    pub callbacks: RefCell<HashMap<String, Rc<Callback>>>,
+}
+
+#[glib::object_subclass]
+impl ObjectSubclass for BuilderRustScope {
+    const NAME: &'static str = "GtkBuilderRustScope";
+    type Type = super::BuilderRustScope;
+    type ParentType = glib::Object;
+    type Interfaces = (BuilderScope,);
+}
+
+impl ObjectImpl for BuilderRustScope {}
+
+impl BuilderScopeImpl for BuilderRustScope {
+    fn create_closure(
+        &self,
+        _builder_scope: &Self::Type,
+        builder: &Builder,
+        function_name: &str,
+        flags: BuilderClosureFlags,
+        object: Option<&glib::Object>,
+    ) -> Result<Closure, glib::Error> {
+        self.callbacks
+            .borrow()
+            .get(function_name)
+            .ok_or_else(|| {
+                glib::Error::new(
+                    BuilderError::InvalidFunction,
+                    &format!("No function named `{}`", function_name),
+                )
+            })
+            .map(|callback| {
+                let callback = callback.clone();
+                let swapped = flags.contains(BuilderClosureFlags::SWAPPED);
+                let object = object.cloned().or_else(|| builder.current_object());
+                if let Some(object) = object {
+                    let object_weak = object.downgrade();
+                    let closure = if swapped {
+                        Closure::new_local(move |args| {
+                            let mut args = args.to_owned();
+                            let object = object_weak
+                                .upgrade()
+                                .expect("Object destroyed before closure invalidated");
+                            let obj_v = unsafe {
+                                let mut v = glib::Value::uninitialized();
+                                glib::gobject_ffi::g_value_init(
+                                    v.to_glib_none_mut().0,
+                                    object.type_().into_glib(),
+                                );
+                                glib::gobject_ffi::g_value_set_object(
+                                    v.to_glib_none_mut().0,
+                                    object.as_object_ref().to_glib_none().0,
+                                );
+                                v
+                            };
+                            args.push(obj_v);
+                            let len = args.len();
+                            args.swap(0, len - 1);
+                            callback(&args)
```

Comment:
For the arguments it seems fine as those are checked on our side, but for the return value is there anything along the path that makes sure that if our callback returns a wrongly typed value it doesn't lead to memory unsafety?

Reply:
See https://github.com/gtk-rs/gtk-rs-core/pull/366#discussion_r753680870 . The correct return value type can actually be checked from inside the `glib::Closure` bindings. And as said, the arguments are correctly checked already.

So I think we're actually fine here in this regard once `glib::Closure` is fixed accordingly.

Reply:
Now that https://github.com/gtk-rs/gtk-rs-core/pull/366 is merged, I did some testing with incorrect value types and it seems to be catching them. I can add some unit tests for it.

Reply:
Thanks for testing!